### ISSUE TITLE
Problem with addMonths function

### ISSUE THREAD

**Issue by magnusottosson** [date-fns/date-fns#3643]
There is an issue with the [addMonths function](https://date-fns.org/v3.0.6/docs/addMonths)

  ## The problem

Adding one month to `2023-03-01T00:00:00.000Z` returns the same month.

  ### 💻 Code demonstrating the problem

  ```js
const result = addMonths(new Date('2023-03-01T00:00:00.000Z'), 1).toISOString()
//=> 2023-03-31T23:00:00.000Z
  ```

  ### 🙁 Actual behavior

Returns end of march: `2023-03-31T23:00:00.000Z`

  ### 🙂 Expected behavior

Should return 1st of april: `2023-04-01T00:00:00.000Z`

It works for other months: 

 ```js
const result = addMonths(new Date('2023-04-01T00:00:00.000Z'), 1).toISOString()
//=> 2023-05-01T00:00:00.000Z
```

  ## Debug information

  - date-fns version: 3.0.6
  - Browser/Node.js version: Node v20.9.0
  - Your timezone: UTC
  - Your current time: 13:55

**Comment by ifrim** [comment]
I think it is because the daylight savings time change is not taken into account.
In the screenshot you can notice that in my timezone the original date is in Eastern European Standard Time, but when you add a month the new date is in Eastern European Summer Time. So an hour was lost on the way.
![add-months-timezone-issue](https://github.com/date-fns/date-fns/assets/5877494/635fdcc2-23d1-41c0-8681-178f5ba72601)

fyi, the same problem exists for other functions like addDays, addWeeks

**Comment by kenlyon** [comment]
@ifrim I think your example of the local date seems to be behaving correctly, though. The local time switches between time zones, so you can't really have a March 1st in Eastern European Summer Time or an April 1st in Eastern European Standard Time.

Where I think it breaks down is when dealing with UTC. I had a test case that was subtracting 25 days from  `"2024-03-25T00:00:00.000Z"`. I expected to get back `'2024-02-29T00:00:00.000Z'` but got `'2024-02-29T01:00:00.000Z'` instead.

**Comment by ifrim** [comment]
@kenlyon yes, so I think that is the problem. All calculations are done in the timezone of where the code gets executed but sometimes you want a specific timezone regardless of where the code happens to execute. So adding one day in a timezone gives a different date than adding it in another timezone.
What I think is needed is a way to specify the timezone in which the calculations should take place.

**Comment by kenlyon** [comment]
@ifrim I think the problem is that you can't really specify one way or the other. A `Date` object always has a local time zone so you can't know just by looking at it whether it should behave as local or UTC. It's going to need to do the same thing in both cases. It's maybe up to us as the consumers to correct it otherwise.

I think in my case when I'm dealing with UTC, I might just do the following steps:

1. Store the initial values of `date.getUTCHours()` and `date.getUTCMinutes()` in variables.
2. Perform the `addDays()`, `addWeeks()` or `AddMonths()` operation.
3. Store the new values of `date.getUTCHours()` and `date.getUTCMinutes()` in other variables.
4. If the hours and minutes have changed (I include minutes in honour of the Noufoundland Time Zone.) use `addHours()` and `addMinutes()` to shift them back in the opposite direction.

Two important points to this:
1. I'm detecting differences rather than hard-coding any magic values. I think this is important because daylight savings happens on different days in different parts of the world, and even for a given time zone, it may even have changed historically to fall on a different date. (I think the US used to do it before Halloween then they changed it.. )
2. I don't want to just set the time to the same time as before, I want to add/subtract hours. The daylight savings change may have caused it to roll into the adjacent day. Overwriting the time would not fix the day/month part.

I think this should be a solid approach.

**Comment by kenlyon** [comment]
I wonder if `addYears()` is affected by leap years.

**Comment by ifrim** [comment]
@kenlyon I don't think that is accurate. A date is just a number, the number of milliseconds from january 1st 1970 UTC. The date object doesn't have a local time zone in itself. The environment in which the code executes has a time zone, and this is used when turning that date into a string.
And if I have to write code when adding dates then what is the point of using a library.
I still think that we should be able to do this by specification/configuration and not by writing code in user land.
But unfortunately I had to do this for one of my use case too.

**Comment by kenlyon** [comment]
@ifrim I agree if you're dealing with a `number`. It's open to interpretation. However, a `Date` has an awareness of "the local time zone". For example, when I ran `new Date()` in my browser console, here was the response:

![image](https://github.com/date-fns/date-fns/assets/10090224/fb6d4bc3-e312-4a08-938f-baa423e01cfe)
I didn't say which time zone to use, but it knows which time zone I am in, based on the browser settings I think.

**Comment by ifrim** [comment]
@kenlyon As I mentioned in my last comment the environment knows the local time zone, the date object isn't "aware" in any way of that time zone.
Taken directly form the spec:
"An ECMAScript time value [is a Number](https://262.ecma-international.org/#sec-ecmascript-language-types-number-type), either a [finite](https://262.ecma-international.org/#finite) [integral Number](https://262.ecma-international.org/#integral-number) representing an instant in time to millisecond precision or NaN representing no specific instant."

**Comment by kenlyon** [comment]
@ifrim We're not talking about the same thing. You're quoting about an ECMAScript time value (which I meant when I said "If you're dealing with a number. I'm talking about a JavaScript [Date](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date) object which contains a method called [getTimezoneOffset()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/getTimezoneOffset). It knows what timezone it should consider "Local".

My initial point was that this still doesn't help with regard to knowing if a given `Date` is to be treated as UTC or local. I can't expect a dateFns function like `addDays()` to take my `Date` and know whether I want midnight local time or midnight UTC.

On a side note, I realize my reference to `minutes` is a moot point. While the Newfoundland timezone is half an hour offset from others, it still changes by an hour for daylight savings so a workaround would only need to inspect the hours.

**Comment by kenlyon** [comment]
Here's what I have so far for my own variation of "addDays" that ensures the time is corrected. It's still a bit limited as it assumes the time starts as midnight.
```
private addDaysSafe(date: Date, amount: number, utc: boolean) {
    if (amount === 0) {
        return date;
    }

    if (utc) {
        const hours = date.getUTCHours();
        date = addDays(date, amount);
        const newHours = date.getUTCHours();
        if (newHours !== hours) {
            const correction = newHours === 23 ? 1 : -1;
            date = addHours(date, correction);
        }
    } else {
        date = addDays(date, amount);
    }

    return date;
}
```

**Comment by kenlyon** [comment]
I just got back to working on this and discovered a much nicer solution. If you're using `date-fns` v3 or later, you can pass in an instance of `UTCDate` ([@date-fns/utc](https://www.npmjs.com/package/@date-fns/utc)) to functions like `addMonths` and all of the calculations will work.

My work on my "safe" versions was working for adding days, but I hit a problem in `addMonths` because I realized `date-fns` uses local date methods like `.getDate()` internally, which was giving the wrong results when I wanted to use UTC.

The `UTCDate` class redirects all local date get and set methods to the UTC equivalents, ensuring the UTC values are being used.

I'm now rewriting my own code to use `UTCDate` when the calculation is to be in UTC, converting back to a regular `Date` before returning the result.
